SlashNext: Submitting SlashNextPhishingIR application to be reviewed …

SaadatAbid opened the pull request:
…by Demisto team.

Signed-off-by: Saadat Abid <saadatabid@gmail.com>

<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Status
Ready

## Related Issues
None. Initial Submission.

## Description
SlashNextPhishingIR application

## Screenshots
N/A

## Related PRs
N/A

## Required version of Demisto
4.5.0

## Does it break backward compatibility?
   - No

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [x] Code Review

## Dependencies
Mention the dependencies of the entity you changed as given from the precommit hooks in checkboxes, and tick after tested them.
- [ ] Dependency 1
- [ ] Dependency 2
- [ ] Dependency 3

## Additional changes
Describe additional changes done, for example adding a function to common server.

## Technical writer review
Mention and link to the files that require a technical writer review.
- [x] [YAML file](link)
- [ ] [CHANGELOG](link)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=4420) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=4420) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=4420) it.</sub>

Itay4 commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+        snx_ioc_cont, dbot_score_cont, url_cont = get_snx_url_ioc_context(url_data, is_scan=True)
+
+        ec = {
+            'SlashNext.IoC(val.Value == obj.Value)': snx_ioc_cont[0],
```
```suggestion
            'SlashNext.IoC(val.Value === obj.Value)': snx_ioc_cont[0],
```

SaadatAbid replied:
OK

Itay4 replied:
Please have a look [here](https://github.com/demisto/content/tree/master/docs/DT)

Itay4 commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+        title = 'SlashNext Phishing Incident Response - URL Scan\n'\
+                '##### url = {}'.format(url_data.get('url'))
+
+        if response.get('normalizeData').get('normalizeStatus') == 1:
```
```suggestion
        if response.get('normalizeData', {}).get('normalizeStatus') == 1:
```

SaadatAbid replied:
Mandatory field

Itay4 replied:
Can you guarantee this convention will stay as is and backward compatibility will be kept forever?

Itay4 commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+                '##### url = {}'.format(url_data.get('url'))
+
+        if response.get('normalizeData').get('normalizeStatus') == 1:
+            title = title + ' *\n*' + response.get('normalizeData').get('normalizeMessage')
```
```suggestion
            title += ' *\n*' + response.get('normalizeData', {}).get('normalizeMessage')
```

SaadatAbid replied:
Ok

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.yml:
```diff
@@ -0,0 +1,1440 @@
+name: SlashNextPhishingIR
+display: SlashNext Phishing Incident Response
+category: Data Enrichment & Threat Intelligence
+image: data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAdkAAAE2CAYAAAAtVehVAAAAAXNSR0IArs4c6QAAQABJREFUeAHsfQlgFFW2dlV1FkLYkVVFHfW5ACKYBEFRUGDcABUIIqs4gsriG8dxdEZnMm7/jKMzPhZZVEB2EjdExBUYRsQsoIggiqOoyCKyB0KS7qr/O7e6kk6nl6ruqk535xRUqurec88997vV9dW5W8kSb4wAIxCXCPReq6WUbSvpo6raIE2Wekia1A6GtpRk6RdZkvbg+iOXLK84s03WfwpyZU9cFoKNYgTqOQL4rfLGCDAC8YSApmlyzvTiXE2TnpQk7dywtsnSl5KkPFQyOfvNsLIswAgwAjFFgEk2pnBzZo ...
```
Ok. So we should remove the image tag altogether from this?

Itay4 replied:
Yes.

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.yml:
```diff
@@ -0,0 +1,1440 @@
+  and more. Automating URL analysis can save IR teams hundreds of hours versus manually
+  triaging these emails or checking URLs and domains against less accurate phishing
+  databases and domain reputation services.  '
+detaileddescription: "The SlashNext Phishing Incident Response Integration app allows
```
Understood.

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.yml:
```diff
@@ -0,0 +1,1440 @@
+  get definitive, binary verdicts (malicious or benign) along with IOCs, screen shots,
+  and more. Automating URL analysis can save IR teams hundreds of hours versus manually
+  triaging these emails or checking URLs and domains against less accurate phishing
+  databases and domain reputation services.  '
```
Ok

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.yml:
```diff
@@ -0,0 +1,1440 @@
+commonfields:
+  id: SlashNextPhishingIR
+  version: -1
+name: SlashNextPhishingIR
```
Ok

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.yml:
```diff
@@ -0,0 +1,1440 @@
+commonfields:
+  id: SlashNextPhishingIR
```
Ok. We'll update this.

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.yml:
Ok

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.yml:
We don't have any. Can you specify some use cases for these? Thanks

Itay4 replied:
These are general integration parameters we use almost in every integration.
So `insecure` is passed to `requests` as `verify`
`proxy` is used to enabled system proxy settings

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.yml:
Ok

SaadatAbid commented on Integrations/SlashNextPhishingIR/CHANGELOG.md:
```diff
@@ -0,0 +1,4 @@
+## [Unreleased]
```
So should I remove the file altogether or just empty the file?

Itay4 replied:
You can remove it

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+AUTH_KEY = demisto.params().get('apikey')
+
+BASE_API = demisto.params().get('apiurl')
+if BASE_API is None:
```
'apiurl' is kept as an optional parameter.

Itay4 replied:
Currently it is set as `required: true`
How can it be optional ?

SaadatAbid replied:
That `required: true` is for `apikey`

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+    return_error(e.message)
+
+finally:
+    LOG('Command execution completed.')
```
Ok

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+except Exception, e:
+    LOG(e)
+    LOG.print_log(False)
+    return_error(e.message)
```
Ok

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+    demisto.results('Elapsed Time = {}ms'.format(end_time - start_time))
+
+except Exception, e:
+    LOG(e)
```
Ok

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+        download_html_command()
+    elif demisto.command() == 'snx-download-text':
+        download_text_command()
+    elif demisto.command() == 'snx-api-quota':
```
Its implemented as to what we want to shown the customers

Itay4 replied:
We prefer not to have this kind of implantation for integration commands.

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+#
+# Copyright (C) SlashNext, Inc. (www.slashnext.com)
```
OK

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+        api_quota_command()
+
+    end_time = int(round(time.time() * 1000))
+    demisto.results('Elapsed Time = {}ms'.format(end_time - start_time))
```
Ok

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+    return snx_ioc_cont_list, dbot_score_cont_list, url_cont_list
+
+
+def retrieve_and_publish_host_repute(host, tag):
```
This has been done to minimize the redundant code as `retrieve_and_publish_host_repute` is being used in multiple commands.

Itay4 replied:
I understand, but it is not according to our convention.

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+    """
+    args = demisto.args()
+
+    if 'host' not in args:
```
Will update accordingly

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+        return_error('Unsupported HTTP method selected for API execution!')
+
+@logger
+def publish_md_and_ec(markdown, snx_contents, entry_context):
```
I had not tried this, Will update if it produces the similar results.

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+        return_error('Missing required command argument "host", Please provide "host" and retry!')
+
+    response = retrieve_and_publish_host_repute(host=args.get('host'), tag='Host Report')
+    if response.get('threatData').get('verdict').startswith('Unrated'):
```
These are mandatory fields so defaults aren't needed

Itay4 replied:
Can you guarantee this convention will stay as is and backward compatibility will be kept forever?

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+
+    response = retrieve_and_publish_host_repute(host=args.get('host'), tag='Host Report')
+    if response.get('threatData').get('verdict').startswith('Unrated'):
+        return
```
`retrieve_and_publish_host_repute` do that. It is intended this way. REQUIREMENTS!  :)

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+
+    # Host Report Section
+    response = retrieve_and_publish_host_urls(host=args.get('host'), tag="Host's Latest Scanned URL", rpp=1)
+    url_data = response.get('urlDataList')[0]
```
These are mandatory fields and never empty in the response so those checks are being skipped

Itay4 replied:
Can you guarantee this convention will stay as is and backward compatibility will be kept forever?

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+    return response
+
+
+def download_forensics_data(scanid, tag, sc=False, html=False, txt=False):
```
Ok

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+
+    # Show Error Message
+    if show_error_msg is True and (sc is True or html is True or txt is True):
+        demisto.results('API Returned, "{}:{}"!'.format(error_no, error_msg))
```
This is a requirement and indicated condition can happen in case of positive results so it's a notification for user instead of an error.

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+        else:
+            show_error_msg = False
+
+            sc_base64 = response.get('scData').get('scBase64')
```
Mandatory field

Itay4 replied:
Can you guarantee this convention will stay as is and backward compatibility will be kept forever?

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+''' EXECUTION '''
+
+
+LOG('Command to be executed is %s.' % (demisto.command(),))
```
OK

SaadatAbid commented on Integrations/SlashNextPhishingIR/SlashNextPhishingIR.py:
```diff
@@ -0,0 +1,976 @@
+''' HELPERS FUNCTIONS '''
+
+@logger
+def invoke_snx_api(api_type='validation', method='POST', **kwargs):
```
This has been done to achieve minimize the code redundancy as per specific requirements of our integration.

Itay4 replied:
I understand, but it is not according to our convention.

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.py:
```diff
@@ -0,0 +1,941 @@
+    except Exception as e:
+        return_error(str(e))
+
+    finally:
```
You can remove this, it is handled by `LOG`

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.yml:
```diff
@@ -0,0 +1,445 @@
+  defaultvalue: ""
+  type: 4
+  required: true
+- display: SlashNext API Base URL
```
URL parameter should be the top one, above the API key

SaadatAbid commented:
Hi Atay,

1- After making the code changes as you suggested, I am seeing issues with `tableToMarkdown` utility function will is giving following error.
`a bytes-like object is required, not 'str'`

2- Also using the function `return_ouputs` is giving malformed display result like appending `b'` with strings and `^^^` as prefix.

3- Also I have a question about commands prefix, Is there a specific reason that we can't use `snx` and we have to use `slashnext`?

Itay4 commented:
> Hi Atay,
> 
> 1- After making the code changes as you suggested, I am seeing issues with `tableToMarkdown` utility function will is giving following error.
> `a bytes-like object is required, not 'str'`
> 
> 2- Also using the function `return_ouputs` is giving malformed display result like appending `b'` with strings and `^^^` as prefix.
> 
> 3- Also I have a question about commands prefix, Is there a specific reason that we can't use `snx` and we have to use `slashnext`?

1. what are you passing to `tableToMarkdown` ?
2. its a bug we fixed
3. `snx` is not indicative enough, we prefer the command prefix to be straight forward and understood

SaadatAbid commented:
I am 

> > Hi Atay,
> > 1- After making the code changes as you suggested, I am seeing issues with `tableToMarkdown` utility function will is giving following error.
> > `a bytes-like object is required, not 'str'`
> > 2- Also using the function `return_ouputs` is giving malformed display result like appending `b'` with strings and `^^^` as prefix.
> > 3- Also I have a question about commands prefix, Is there a specific reason that we can't use `snx` and we have to use `slashnext`?
> 
>     1. what are you passing to `tableToMarkdown` ?
> 
>     2. its a bug we fixed
> 
>     3. `snx` is not indicative enough, we prefer the command prefix to be straight forward and understood

1- I am using a string for the name like this `'SlashNext Phishing Incident Response - {}\n'\
                '##### host = {}'.format(tag, host)` but I see it working at one place and not another.

2- OK

3- OK

Itay4 commented:
@SaadatAbid great, will review soon
in the mean time, please note that the [CircleCI build](https://circleci.com/gh/demisto/content/30245) fails

SaadatAbid commented:
> @SaadatAbid great, will review soon
> in the mean time, please note that the [CircleCI build](https://circleci.com/gh/demisto/content/30245) fails
> also, please pull from base branch

@Itay4 Please have a look at the CircleCI build I think we need to do some whitelisting for SlashNext
also, base branch is pulled. In addition to that, 
1- Integration documentation is generated,
2- Test playbook is developed,
3- Sample unit test is developed for your feedback.

Please have a quick review. Thanks.

Itay4 commented:
Yes, please add the entities mentioned in the CircleCI build to [the whitelist](https://github.com/demisto/content/blob/master/Tests/secrets_white_list.json)
Also, please remove `™` from the integration description
Will review soon
Would also like to schedule an execution review, can you please reach out to me on DFIR?

SaadatAbid commented:
>     1. Yes, please add the entities mentioned in the CircleCI build to [the whitelist](https://github.com/demisto/content/blob/master/Tests/secrets_white_list.json)
> 
>     2. Also, please remove `™` from the integration description
> 
>     3. The linting step is also failing, please add the following lines to the top of the Python code:
> 
> 
> ```python
> import demistomock as demisto
> from CommonServerPython import *
> from CommonServerUserPython import *
> ```
> 
>     1. Will review soon
> 
>     2. Would also like to schedule an execution review, can you please reach out to me on DFIR?

How can I reach you on DFIR? BTW what is DFIR?

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.py:
```diff
@@ -0,0 +1,975 @@
+        data['authkey'] = AUTH_KEY
+        response = requests.request(method, url=url, params=data, data=data, timeout=300)
+        if response.status_code == 200:
+            return response.json()
```
this might raise `JSONDecodeError`, please wrap it in `try-except` block

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.py:
```diff
@@ -0,0 +1,975 @@
+    :param method: HTTP method to be used for API i.e. GET or POST
+    :return: Response of the SlashNext web API in json format
+    """
+    if method == 'GET' or method == 'POST':
```
this condition can be removed, it is not needed

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.py:
```diff
@@ -0,0 +1,975 @@
+        response = None
+        url = BASE_API + endpoint
+        data['authkey'] = AUTH_KEY
+        response = requests.request(method, url=url, params=data, data=data, timeout=300)
```
is it needed to pass `data` to `params` as well ?

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.py:
```diff
@@ -0,0 +1,975 @@
+
+''' GLOBAL VARS '''
+AUTH_KEY = demisto.params().get('apikey')
+BASE_API = demisto.params().get('apiurl', 'https://oti.slashnext.cloud/api')
```
need to handle the case in which there's a trailing slash in the URL

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.yml:
```diff
@@ -0,0 +1,445 @@
+  type: 4
+  required: true
+script:
+  dockerimage: demisto/python3:3.7.3.221
```
```suggestion
  dockerimage: demisto/python3:3.7.3.260
```

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.yml:
```diff
@@ -0,0 +1,445 @@
+  name: apiurl
+  defaultvalue: https://oti.slashnext.cloud/api
+  type: 0
+  required: false
```
this should be required

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.py:
```diff
@@ -0,0 +1,975 @@
+
+    else:
+        dbot_landing_score_cont, landing_url_cont = get_dbot_std_context(
+            url_data.get('landingUrl').get('url'), 'url',
```
no need to get `landingUrl` and `threatData` every time, do it once and then get the fields you want out of it

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.py:
```diff
@@ -0,0 +1,975 @@
+
+
+def main():
+    LOG('Command to be executed is %s.' % (demisto.command(),))
```
don't use `%s`, use f-strings or strings format

SaadatAbid commented:
> Looks good !
> A few final comments:
> 
>     1. Add `DBotScore` to the outputs in the YAML
> 
>     2. Change the context path from `SlashNext.IoC` to the actual IOC type, like `SlashNext.URL`
> 
>     3. Split each command function in to two functions, the one that handles all the logic, like extract arguments and parsing outputs, and one that calls `http_request`. The command function should call the request one.
> 
>     4. Can we add general IOC reputation commands, like `ip`, `domain`, `url` ?

1- OK
2- This SlashNext.IoC design is intentional to keep the structure more generic like DBotScore and  also there is a Type key for the type of IoC.
3- Can you elaborate?
4- Do you mean like divide the slashnext-host-reputation into slashnext-ip-reputation and slashnext-domain-reputation

Itay4 commented:
2. it is better to split them by IOC type, so it will be easier to reach to it in a playbook
3. please have a look [here](https://github.com/demisto/content/tree/master/docs/code_conventions#command-functions)
4. no. for TI integration we have generic IOC commands (`ip`, `domain`, etc..), which ease the process of enrichment. please have look at [Recorded Future integration](https://github.com/demisto/content/tree/master/Integrations/Recorded_Future) for reference

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.py:
```diff
@@ -0,0 +1,1327 @@
+    enc_host = host.encode('idna')
+
+    title = 'SlashNext Phishing Incident Response - Host Report\n'\
+            '##### host = {}'.format(enc_host)
```
No need for new line

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.py:
```diff
@@ -0,0 +1,1327 @@
+    })
+
+
+def api_quota_command():
```
Please remove

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.py:
```diff
@@ -0,0 +1,1327 @@
+
+    if response.get('errorNo') == 1:
+        demisto.results(
+            'Your Url Scan request is submitted to the cloud and may take up-to 60 seconds to complete.\n'
```
Please push the ID to context as well

SaadatAbid commented:
>     1. it is better to split them by IOC type, so it will be easier to reach to it in a playbook
> 
>     2. please have a look [here](https://github.com/demisto/content/tree/master/docs/code_conventions#command-functions)
> 
>     3. no. for TI integration we have generic IOC commands (`ip`, `domain`, etc..), which ease the process of enrichment. please have look at [Recorded Future integration](https://github.com/demisto/content/tree/master/Integrations/Recorded_Future) for reference

@Itay4 
1. The IoC have been split into IP, Domain and URL contexts in SlashNext
2. Done.
3. We have implemented generic commands to ip and domain. We can offer url command at the moment due to a different system behavior at our end.

Finally we have addressed all the comments as per our execution review. I have put forward the request for an API key for you as well. Will send you as soon as I get that.

Itay4 commented:
@kirbles19 please review the [integration YAML file](https://github.com/demisto/content/blob/752dcb9079a80e56f27a410db54d12617c8e5889/Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.yml) and the [detailed instructions](https://github.com/demisto/content/blob/752dcb9079a80e56f27a410db54d12617c8e5889/Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse_description.md)

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.yml:
```diff
@@ -0,0 +1,732 @@
+  and more. Automating URL analysis can save IR teams hundreds of hours versus manually
+  triaging these emails or checking URLs and domains against less accurate phishing
+  databases and domain reputation services."
+configuration:
```
Please add `proxy` and `insecure` parameters

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.py:
```diff
@@ -0,0 +1,1180 @@
+    }
+
+    if domain_cont is not None:
+        ec['Domain'] = domain_cont
```
Should return `Domain` standard context in any case

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.py:
```diff
@@ -0,0 +1,1180 @@
+    }
+
+    if ip_cont is not None:
+        ec['IP'] = ip_cont
```
Should return `IP` standard context in any case

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.py:
```diff
@@ -0,0 +1,1180 @@
+
+    if response.get('errorNo') == 1:
+        ec = {
+            'SlashNext.URL.ScanID': response.get('urlData').get('scanId')
```
Also return here the `URL`
Need to use here DT, otherwise running twice will create two entries in the context

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.py:
```diff
@@ -0,0 +1,1180 @@
+            'Your Url Scan request is submitted to the cloud and may take up-to 60 seconds to complete.\n'
+            'Please check back later using "slashnext-download-html" command with Scan ID = {}'.format(scanid))
+    elif response.get('errorNo') != 0:
+        demisto.results('API Returned, {}:{}!'.format(response.get('errorNo'), response.get('errorMsg')))
```
Should be returned as an error
![image](https://user-images.githubusercontent.com/31018228/66460380-7703e180-ea7f-11e9-830c-8adc67d908fe.png)

Itay4 commented on Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.py:
```diff
@@ -0,0 +1,1180 @@
+    dbot_score_cont = {
+        'Indicator': indicator,
+        'Type': ioc_type.lower(),
+        'Vendor': 'SlashNext',
```
`Vendor` should be the same as the integration `ID`

SaadatAbid commented:
> Some final comments:
> 
>     1. Please remove all exclamation marks from the error messages.
> 
>     2. Please remove the subtitle (e.g. `'##### ip = {}'.format(ip)`) from all tables
> 
>     3. Please also add `url` reputation command, like you did for `ip` and ` domain`
> 
>     4. Almost all commands should return to general context standard, like `IP`, `Domain` and `URL`, please make sure that they do.

1- Ok
2- I know you want us to follow certain conventions but this following this one can cause user more confusion as in case of an API error user won't know from which input actually caused this specially in playbooks where multiple things will be logged to evidence board.   
3- As I mentioned earlier, We do not offer such functionality at this moment due to certain system behavior so we can not add this.
4- This behavior is as per [Demisto documentation](https://github.com/demisto/content/tree/master/docs/dbot) where these context should be added in case of only a Malicious verdict i.e. if DBot score is 3.

Itay4 commented:
regarding point 4, only the `Malicious` object should be added in case of DBotScore 3, but the rest `IP.Address`, `URL.Data`, `Domain.Name` should always be populated

SaadatAbid commented:
> regarding point 4, only the `Malicious` object should be added in case of DBotScore 3, but the rest `IP.Address`, `URL.Data`, `Domain.Name` should always be populated

This has been updated as per your instructions.

Itay4 commented:
@kirbles19 please review the following:
- https://github.com/demisto/content/blob/b6838c899696b9d3a00fd467e159806997445f72/Integrations/SlashNextPhishingIncidentResponse/README.md
 - https://github.com/demisto/content/blob/b6838c899696b9d3a00fd467e159806997445f72/Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse.yml
 - https://github.com/demisto/content/blob/b6838c899696b9d3a00fd467e159806997445f72/Integrations/SlashNextPhishingIncidentResponse/SlashNextPhishingIncidentResponse_description.md

kirbles19 commented:
@Itay4 Done.
